src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,414 @@
+
+        protected override char ClosingBrace => CurlyBrace.CloseCharacter;
+
+        protected override int OpeningTokenKind => (int)SyntaxKind.OpenBraceToken;
+
+        protected override int ClosingTokenKind => (int)SyntaxKind.CloseBraceToken;
+
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            // Get the base brace completion result with the new brace.
+            var baseCompletionResult = await base.GetBraceCompletionAsync(braceCompletionContext, cancellationToken).ConfigureAwait(false);
+            if (baseCompletionResult == null)
+            {
+                return null;
+            }
+
+            var originalText = await braceCompletionContext.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textWithClosingBrace = originalText.WithChanges(baseCompletionResult.Value.TextChange);
+            // Since we inserted a closing brace, the closing point now is one higher.
+            var closingPoint = braceCompletionContext.ClosingPoint + 1;
+            Debug.Assert((textWithClosingBrace)[closingPoint - 1] == ClosingBrace);
+
+            var textWithAllChanges = await FormatTrackingSpanAsync(braceCompletionContext.Document.WithText(textWithClosingBrace), braceCompletionContext.OpeningPoint, closingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: true, cancellationToken).ConfigureAwait(false);
+
+            var caretLocation = GetCloseBraceLocation(braceCompletionContext.OpeningPoint, textWithAllChanges);
+            var textChange = await GetCollapsedTextChangeAsync(textWithAllChanges, braceCompletionContext.Document, cancellationToken).ConfigureAwait(false);
+            return new BraceCompletionResult(textChange, caretLocation);
+
+            int GetCloseBraceLocation(int openingPoint, SourceText text)
+            {
+                var closingParenLocation = text.IndexOf(ClosingBrace.ToString(), openingPoint, caseSensitive: false);
+                // We just added the closing brace so we should never encounter a situation where we can't get the location.
+                Debug.Assert(closingParenLocation != -1);
+
+                return closingParenLocation;
+            }
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext context, CancellationToken cancellationToken, bool supportsVirtualSpace = true)
```

Comment:
this comes from CurlyBraceCompletionSession.AfterStart, with modifications to use documents and supporting non-virtual spaces for the caret location